### PR TITLE
SYNR-1301 check for version update

### PR DESCRIPTION
**Pull request by kimyen**

### CODE REVIEW

File: R/version_check.R [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+# check package_name's version against old.packages() output to see if the
+# package version is out of date upto the given precision
+.isVersionOutOfDate <- function(info, package_name, precision) {
+  if (!is(info, "matrix") || !(package_name %in% rownames(info))) {
+    return(FALSE)
+  }
+  current_version <- .simplifyVersion(info[package_name, "Installed"], precision)
+  latest_version <- .simplifyVersion(info[package_name, "ReposVer"], precision)
+  return(current_version < latest_version)
```
Comment (karawoo) [inline R/version_check.R]: I don't think this line will correctly handle all cases, e.g.: 

``` r
current_version <- "0.5.0"
latest_version <- "0.10.0"
current_version < latest_version
#> [1] FALSE
```

It might be worth using R's built in `package_version` class:

```
package_version(current_version) < package_version(latest_version)
#> [1] TRUE
```

<sup>Created on 2019-02-26 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

File: R/version_check.R [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+.PACKAGE_NAME = "synapser"
+
+# convert the full version to a precision
+# precision can be one of the following values:
+.MAJOR <- 1
+.MINOR <- 2
+.PATCH <- 3
+.simplifyVersion <- function(version, precision) {
+  if (!match(precision, c(.MAJOR, .MINOR, .PATCH))) {
+    throw("Invalid precision: ", precision)
```
Comment (karawoo) [inline R/version_check.R]: If I'm not mistaken, `throw()` is from the package R.methodsS3 so we'd need to take it on as a dependency and write this as `R.methodsS3::throw()` (or just use `stop()`)

File: R/version_check.R (hunk 1 above)
Comment (kimyen) [reply R/version_check.R]: I will add that as a test case to my `testthat` cases. 
Also, I was looking at `compareVersion()`: https://stat.ethz.ch/R-manual/R-devel/library/utils/html/compareVersion.html

File: R/version_check.R [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+.RAN <- "https://sage-bionetworks.github.io/ran"
+.PACKAGE_NAME = "synapser"
+
+# convert the full version to a precision
+# precision can be one of the following values:
+.MAJOR <- 1
+.MINOR <- 2
+.PATCH <- 3
+.simplifyVersion <- function(version, precision) {
+  if (!match(precision, c(.MAJOR, .MINOR, .PATCH))) {
```
[the remark this answers is not in the document]
Comment (kimyen) [reply R/version_check.R]: ahh, it looks like I missed some test cases. 
Also, I used `%in%` in other parts of the code because my test cases caught `match` errors. Thank you for catching this.

File: R/version_check.R (hunk 2 above)
Comment (kimyen) [reply R/version_check.R]: oh, I didn't know that. Thank you!

**Comment by kimyen** [comment]
@karawoo I fixed the bugs that you found. Could you review this code again? Thank you.

**Comment by kimyen** [comment]
This is great! Thanks @karawoo for the links. You pointed out that my code wasn't written according to best practices in R and provide documentation for me to understand why. 

I do agree with using default values in the function signature over using dynamic lookup. However, with precision case, using a number causes lost of readability there. Someone else looks at my "new" code with precision as number would not understand that precision must be 1, 2, or 3, and what 1, 2, and 3 mean explicitly. Ideally, I would like precision to be an enum where the value can only be 1, 2, or 3. However, I don't think enum exist in R. A named list is the first that came to mind: 
```precision <- c(MAJOR = 1, MINOR = 2, PATCH = 3)```
However, this gets me back to using dynamic lookup. Do you have any other recommendations such that I can keep the readability of what `precision` is, and what each valid value means?

**Comment by kimyen** [comment]
How about:
```
.simplifyVersion <- function(version, precision) {
  major <- 1
  minor <- 2
  patch <- 3
  if (!(precision %in% c(major, minor, patch))) {
    stop("Invalid precision: ", precision)
  }
  parts <- unlist(strsplit(version, "[.]"))
  parts[is.na(parts[1:precision])] <- "0"
  return(paste(parts[1:precision], collapse = "."))
}
```

**Comment by karawoo** [comment]
> Someone else looks at my "new" code with precision as number would not understand that precision must be 1, 2, or 3, and what 1, 2, and 3 mean explicitly.

I'd normally document this type of thing in the roxygen comments for the function like so: 

```r
#' @param precision The level of precision to which version should be
#'   simplified. Can be 1, 2, or 3 corresponding to major, minor, or patch
```

but I see synapser is not using roxygen. Defining `major`, `minor`, and `patch` within the function seems fine to me. The named vector option also works (and doesn't have the dynamic lookup problem if you define it within the function):

```r
.simplifyVersion <- function(version, precision) {  
  if (!(precision %in% c(MAJOR = 1, MINOR = 2, PATCH = 3))) {
    stop("Invalid precision: ", precision)
  }
  parts <- unlist(strsplit(version, "[.]"))
  parts[is.na(parts[1:precision])] <- "0"
  return(paste(parts[1:precision], collapse = "."))
}
```

Possibly even easier from the user perspective would be to accept `precision` as a string

```r
.simplifyVersion <- function(version, precision) {
  precision_opts <- c("major", "minor", "patch")
  if (!(precision %in% precision_opts)) {
    stop("Invalid precision: ", precision)
  }
  precision_num <- which(precision_opts == precision)
  parts <- unlist(strsplit(version, "[.]"))
  parts[is.na(parts[1:precision_num])] <- "0"
  return(paste(parts[1:precision_num], collapse = "."))
}
```

I think any of these three versions (yours above, or these two) would be fine.

**Comment by kimyen** [comment]
This is great! Thank you so much @karawoo ! I will make the changes.

**Comment by karawoo** [comment]
Cool! I had another thought -- the message string in `.printVersionOutOfDateWarnings()` has our ran hardcoded in part of the string but not others. It might be good to generalize it to substitute in the package name and ran everywhere these appear

```r
.printVersionOutOfDateWarnings <- function(current_version, latest_version, package, ran) {
  message <- sprintf(
    "\nNew %s version detected: You are using %s version %s. %s version %s is detected at %s. To upgrade to the latest version of synapser, please run the following command: install.packages(\"%s\", repos=\"%s\")\n",
    package,
    package,
    current_version,
    package,
    latest_version,
    ran,
    package,
    ran
  )
  packageStartupMessage(message)
}
```

If it's too hard to follow all the `%s`'s this could use `paste0()` instead.

**Comment by kimyen** [comment]
Yeah, you're right. I missed that.
